packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmSetTests.kt

```diff
@@ -870,6 +973,56 @@ internal abstract class ManagedSetTester<T>(
         }
     }
 
+    override fun iterator_concurrentModification() {
+        val dataSet = typeSafetyManager.dataSetToLoad
+
+        errorCatcher {
+            realm.writeBlocking {
+                val set = typeSafetyManager.createContainerAndGetCollection(this)
+                set.add(dataSet[0])
+
+                // Add something to the set to trigger a ConcurrentModificationException
+                val addIterator = set.iterator()
+                addIterator.next()
+                set.add(dataSet[1])
+                assertFailsWith<ConcurrentModificationException> {
+                    addIterator.remove()
+                }
+
+                // Clear set to avoid issues with datasets of different lengths
+                set.clear()
+                set.add(dataSet[0])
+
+                // Remove something from the set to trigger a ConcurrentModificationException
+                // TODO https://github.com/realm/realm-kotlin/issues/1097
+                //  Ignore RealmObject because we can assess structural equality
+                if (classifier != RealmObject::class) {
+                    val removeIterator = set.iterator()
+                    removeIterator.next()
+                    set.remove(dataSet[0])
+                    assertFailsWith<ConcurrentModificationException> {
+                        removeIterator.remove()
+                    }
+                }
+
+                // Clear set to avoid issues with datasets of different lengths
+                set.clear()
+                set.add(dataSet[0])
+
+                // Clear the set to trigger a ConcurrentModificationException
+                val clearIterator = set.iterator()
+                clearIterator.next()
+                set.clear()
+                assertFailsWith<ConcurrentModificationException> {
+                    clearIterator.remove()
+                }
+            }
+        }
+
+        // Makes no sense to test concurrent modifications outside the transaction, so clean up only
+        assertContainerAndCleanup()
```

Comment:
Funny name if we sometimes don't assert. Should the actual clean up just be somewhere else in the test infrastructure?

Reply:
It can't really be moved to `@AfterTest` because it has to be done after every type is tested inside the same test to avoid overcomplicating the setup process for each type - but of course it could be done differently, perhaps should have been done [here](https://github.com/realm/realm-kotlin/issues/941) but we never prioritised it high enough so 🤷 . In any case, I think it's fine having the assertions as an optional parameter but I grant you the name could be better, but you know, `#naming`...